objax/scheduler/scheduler.py

```diff
@@ -0,0 +1,126 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+__all__ = ['LinearAnnealing', 'StepDecay']
+
+
+from typing import List, Tuple, Union
+
+import jax.numpy as jn
+
+
+class Scheduler:
+    def __init__(self,
```

Comment:
Some high level questions/suggestions about scheduler:

1. I think we discussed in the past that, maybe we want to use floating point `step` or `progress` counter instead of integer step counter. This way user may choose whether to use it as actual step counter, as fractional epoch or whatever. Right now it looks like you using integer step counter everywhere in the code
2. Currently the base class takes three arguments `is_cycle`, `max_step` and `base_lr` to the `__init__` function. However it looks to me that only `base_lr` is generic enough that almost every LR scheduler have it, while `is_cycle` and `max_step` sounds too specific and only used by some schedulers but not others.
3. It seems like scheduler tries to automatically increment step number. I think it's generally bad idea and may lead to confusion and bugs, especially given how Objax tracks training steps in other examples. In particular, you can not call scheduler twice at the same training step, otherwise step counter would increase. On the other hand calling scheduler twice may be convenient when you want to save current LR into summary or print it for debugging

Reply:
Done